pmlc/dialect/pxa/transforms/stencil.cc

```diff
@@ -270,19 +509,46 @@ void Stencil::DoStenciling() {
   CollectUsedIndices();
   CollectStrideOneIndices();
 
+  auto ranges = op.getConstantRanges();
+  if (!ranges)
+    return;
+
+  opConstRanges = *ranges;
+
+  // Search tensors' order, inner index and their tiles
+  SearchTensorsOrder();
+
+  IVLOG(1, "Best Perf: " << bestPerf);
+  IVLOG(1, "Best Tiles: " << bestTiles[0] << ":" << bestTiles[1] << ":"
+                          << bestTiles[2]);
+
+  if (bestPerf == std::numeric_limits<double>::max()) {
+    IVLOG(1, "No tile plan for stencil.");
+    return;
+  }
+
   op.setAttr("is_gemm", mlir::UnitAttr::get(op.getContext()));
 }
 
 void StencilPass::runOnFunction() {
   auto func = getFunction();
   func.walk([&](mlir::AffineParallelOp op) {
-    Stencil as(op);
+    unsigned threads = numThreads.getValue();
+    if (threads == 0) {
+      threads = std::thread::hardware_concurrency();
+    }
+    Stencil as(op, threads);
     as.DoStenciling();
   });
 }
 
-std::unique_ptr<mlir::Pass> createStencilPass() {
-  return std::make_unique<StencilPass>();
+StencilPass::StencilPass(const StencilPassOptions &options) {
+  numThreads.setValue(options.numThreads);
+}
+
+void createStencilPass(mlir::OpPassManager &pm,
```

Comment:
See the other pass creation functions in MLIR (and our code). They don't take in a pass manager, instead they just return a unique pointer to the pass. I think we should be consistent here.

Reply:
In the MLIR code I could find a single pass that uses instance options (all the rest use static ones, which I mentioned up in this review). The pass is TestOptionsPass in TestPassManager.cpp.
From the MLIR doc on instance pass it looks like you need to add the pass to the PM and that would cause the options to be parsed.
The doc is at https://github.com/llvm/llvm-project/blob/master/mlir/docs/WritingAPass.md#instance-specific-pass-options